src/compression.rs

```diff
@@ -1189,13 +1180,63 @@ impl CompressionStepResult {
     }
 }
 
+/// the complete result of a single step of compression, this is a somewhat expensive data structure
+/// to create.
+#[derive(Debug, Clone)]
+pub struct CompressionStepResult {
+    pub inv: Invention,
+    pub done: FinishedPattern,
+    pub dc_inv_str: String,
+    pub train_data: CompressionStepData,
+    pub test_data: Option<CompressionStepData>,
+}
+
+impl CompressionStepResult {
+    fn new(done: FinishedPattern, train_programs_node: Id, test_programs_node: Option<Id>, inv_name: &str, shared: &mut SharedData, past_invs: &Vec<CompressionStepResult>, first_train_cost: Option<i32>, first_test_cost: Option<i32>) -> Self {
```

Comment:
Note that the very first costs are now passed as parameters to this function. This was because it got a bit messy keeping track of them in the `CompressionStepData`s... My brain is a bit fried, so I don't remember the exact issue, but I think this felt like the most sane solution to me (at the time at least).